packages/core/database/src/utils/identifiers/index.ts

```diff
@@ -21,25 +41,164 @@ type NameOptions = {
   maxLength?: number;
 };
 
-export const getNameFromTokens = (tokens: NameToken[], max: number = MAX_DB_IDENTIFIER_LENGTH) => {
-  const fullLength = tokens
+// returns a hash of length len
+export function createHash(data: string, len: number): string {
+  if (!_.isInteger(len) || len <= 0) {
+    throw new Error(`createHash length must be a positive integer, received ${len}`);
+  }
+
+  // TODO: shake256 is based on SHA-3 and is slow, we don't care about cryptographically secure, only uniqueness and speed
+  //       investigate alternatives before releasing this. But it is only run on startup, so it should be fine.
+  const hash = crypto.createHash('shake256', { outputLength: Math.ceil(len / 2) }).update(data);
+  return hash.digest('hex').substring(0, len);
+}
+
+export function tokenWithHash(name: string, len: number) {
+  if (!_.isInteger(len) || len <= 0) {
+    throw new Error(`tokenWithHash length must be a positive integer, received ${len}`);
+  }
+  if (name.length <= len) {
+    return name;
+  }
+  if (len < MIN_TOKEN_LENGTH + HASH_LENGTH) {
+    throw new Error(
+      `length for part of identifier too short, minimum is hash length (${HASH_LENGTH}) plus min token length (${MIN_TOKEN_LENGTH}), received ${len} for token ${name}`
+    );
+  }
+
+  const availableLength = len - HASH_LENGTH - HASH_SEPARATOR.length;
+  if (availableLength < MIN_TOKEN_LENGTH) {
+    throw new Error(
+      `length for part of identifier minimum is less than min token length (${MIN_TOKEN_LENGTH}), received ${len} for token ${name}`
+    );
+  }
+
+  return `${name.substring(0, availableLength)}${HASH_SEPARATOR}${createHash(name, HASH_LENGTH)}`;
+}
+
+export function getNameFromTokens(nameTokens: NameToken[], maxLength = MAX_DB_IDENTIFIER_LENGTH) {
+  if (!_.isInteger(maxLength) || maxLength < 0) {
+    throw new Error('maxLength must be a positive integer or 0 to disable');
+  }
+
+  // Ensure all tokens are in snake_case
+  nameTokens.forEach((token) => {
+    if (token.name !== _.snakeCase(token.name)) {
+      throw new Error(`all names must already be in snake_case; received ${token.name}`);
+    }
+  });
+
+  const fullLengthName = nameTokens.map((token) => token.name).join(IDENTIFIER_SEPARATOR);
+
+  // if it fits, or maxLength is disabled, return full length string
+  if (fullLengthName.length <= maxLength || maxLength === 0) {
+    return fullLengthName;
+  }
+
+  // Split tokens by compressibility
+  const { compressible, incompressible } = nameTokens.reduce(
+    (acc: { compressible: NameToken[]; incompressible: NameToken[] }, token) => {
```

Comment:
```suggestion
  const { compressible, incompressible } = nameTokens.reduce<{ compressible: NameToken[]; incompressible: NameToken[] }>(
    (acc, token) => {
```

not important, but just incase you didnt know this lol